e2e/pos_test.go

```diff
@@ -0,0 +1,617 @@
+}
+
+// getBigDefaultStakedBalance returns the default staked balance as a *big.Int
+func getBigDefaultStakedBalance(t *testing.T) *big.Int {
+	val := stakingHelper.DefaultStakedBalance
+	bigDefaultStakedBalance, err := types.ParseUint256orHex(&val)
+	if err != nil {
+		t.Fatalf("unable to parse DefaultStakedBalance, %v", err)
+	}
+
+	return bigDefaultStakedBalance
+}
+
+// validateValidatorSet makes sure that the address is present / not present in the
+// validator set, as well as if the validator set is of a certain size
+func validateValidatorSet(
+	address types.Address,
+	client *jsonrpc.Client,
+	t *testing.T,
+	expectedExistence bool,
+	expectedSize int,
+) {
+	validatorSet, validatorSetErr := framework.GetValidatorSet(address, client)
+	if validatorSetErr != nil {
+		t.Fatalf("Unable to fetch validator set, %v", validatorSetErr)
+	}
+	assert.NotNil(t, validatorSet)
+	assert.Len(t, validatorSet, expectedSize)
+
+	if expectedExistence {
+		assert.Truef(
+			t,
+			foundInValidatorSet(validatorSet, address),
+			"expected address to be present in the validator set",
+		)
+	} else {
+		assert.Falsef(t,
+			foundInValidatorSet(validatorSet, address),
+			"expected address to not be present in the validator set",
+		)
+	}
+}
+
+func TestPoS_Stake(t *testing.T) {
+	stakerKey, stakerAddr := tests.GenerateKeyAndAddr(t)
+	defaultBalance := framework.EthToWei(100)
+	stakeAmount := framework.EthToWei(5)
+
+	numGenesisValidators := IBFTMinNodes
+	ibftManager := framework.NewIBFTServersManager(
+		t,
+		numGenesisValidators,
+		IBFTDirPrefix,
+		func(i int, config *framework.TestServerConfig) {
+			config.SetSeal(true)
+			config.SetEpochSize(2) // Need to leave room for the endblock
+			config.PremineValidatorBalance(defaultBalance)
+			config.Premine(stakerAddr, defaultBalance)
+			config.SetIBFTPoS(true)
+		})
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	ibftManager.StartServers(ctx)
+
+	srv := ibftManager.GetServer(0)
+
+	client := srv.JSONRPC()
+
+	// Stake Balance
+	stakeError := framework.StakeAmount(
+		stakerAddr,
+		stakerKey,
+		stakeAmount,
+		srv,
+	)
+	if stakeError != nil {
+		t.Fatalf("Unable to stake amount, %v", stakeError)
+	}
+
+	// Check validator set
+	validateValidatorSet(stakerAddr, client, t, true, numGenesisValidators+1)
+
+	// Check the SC balance
+	bigDefaultStakedBalance := getBigDefaultStakedBalance(t)
+
+	scBalance := framework.GetAccountBalance(staking.AddrStakingContract, client, t)
+	expectedBalance := big.NewInt(0).Mul(
+		bigDefaultStakedBalance,
+		big.NewInt(int64(numGenesisValidators)),
+	)
+	expectedBalance.Add(expectedBalance, stakeAmount)
+
+	assert.Equal(t, expectedBalance.String(), scBalance.String())
+
+	stakedAmount, stakedAmountErr := framework.GetStakedAmount(stakerAddr, client)
+	if stakedAmountErr != nil {
+		t.Fatalf("Unable to get staked amount, %v", stakedAmountErr)
+	}
+
+	assert.Equal(t, expectedBalance.String(), stakedAmount.String())
```

Comment:
Was your intention to check the total amount that is staked in the smart contract or the amount that this stakerAddr has staked?

Because it does the first, which was already checked in the previous assert if I'm not mistaken.

Reply:
You're right - we can only check the total staked amount on the Staking SC.

The first check asserts that the account (address) balance is correct for the staking address.
The second check asserts that the `stakedAmount` value is set correctly on the contract itself.

These two values need to be the same